projeuler22.rb

```diff
@@ -1,59 +1,47 @@
 # Using names.txt (right click and 'Save Link/Target As...'), a 46K text file containing over five-thousand first names, begin by sorting it into alphabetical order.
 # Then working out the alphabetical value for each name, multiply this value by its alphabetical position in the list to obtain a name score.
-# For example, when the list is sorted into alphabetical order, COLIN which is worth 3 + 15 + 12 + 9 + 14 = 53, is the 938th name in the list. So, COLIN would obtain a score of 938 × 53 = 49714.
+# For example, when the list is sorted into alphabetical order, COLIN which is worth 3 + 15 + 12 + 9 + 14 = 53, is the 938th name in the list.
+#So, COLIN would obtain a score of 938 × 53 = 49714.
 # What is the total of all the name scores in the file?
 require "pry"
 
-alphabet_arr = ("A".."Z").to_a
-n = 0
-HASH_CONVERSION = {}
-alphabet_arr.each do |l|
-  n += 1
-  HASH_CONVERSION[l] = n
-end
-
-module CONVERT_CHARA
-  def convert_name(name)
-    char_arr = name.split(//).to_a
-    @num_value = []
-    char_arr.each do |char|
-      @num_value << HASH_CONVERSION[char]
-    end
-    @num_sum = @num_value.sum
-  end
-end
-
 class NameCalculator
   attr_reader :namelist
   def initialize(namelist)
     @namelist = File.read(namelist).scan(/\w+/).sort
   end
 
-  def name_scores
-    extend CONVERT_CHARA
-    @alpha_value = 0
-    @name_scores = {}
-    @namelist.each do |name|
-      @alpha_value += 1
-      convert_name(name)
-      @name_scores[name] = @alpha_value * @num_sum
+  def convert_name(name) #convert name to upcase chars array, map using alphabet hash to convert chars to numerical score, return the summed value of the char values
+    alphabet_num_hash = ( ( "A".."Z" ).to_a.zip( ( 1..26 ).to_a) ).to_h
```

Comment:
There are 5000 names in the list
That means that this hash will be generated 5000 times

Take a look at the time to completion before and after this change
This is a great use of an instance variable!

You could define a method called `alphabet_num_hash` to return this hash

In the method, we could make it so that it calculates this hash _once_ and if there's already a hash, just return the one that already exists in that `NameCalculator` instance

```ruby
def alphabet_num_hash
  if @alphabet_num_hash # Check for the presence of that instance var
    @alphabet_num_hash # If it exists, implicitly return that instance var
  else # Otherwise, set it
    @alphabet_num_hash = ( ( "A".."Z" ).to_a.zip( ( 1..26 ).to_a) ).to_h
  end
end
```

Now that will only be calculated once!

But we can tidy that syntax with the `||` (or) operator

```ruby
def alphabet_num_hash
  @alphabet_num_hash ||= ( ( "A".."Z" ).to_a.zip( ( 1..26 ).to_a) ).to_h
end
```

This process is called [_memoization_](https://www.justinweiss.com/articles/4-simple-memoization-patterns-in-ruby-and-one-gem/)